### PR TITLE
refactor(slider): Functional slider tick visuals with css background

### LINKED ISSUE #4729: [mdc-slider] Replace cumbersome track marker (discrete tick marks) logic with simple background style ruleset
Not so much a bug, as an optimization...

You can remove most/all the logic of your `appendTrackMarkers`, `removeTrackMarkers`, & `setLastTrackMarkersStyleProperty` code by simply using background style like this...

```css
.mdc-slider--discrete.mdc-slider--display-markers .mdc-slider__track-marker-container {
  --mdc-slider-tick-count: 50;
  --mdc-slider-tick-width: 2px;
  background:
    linear-gradient(
      to right,
      var(--mdc-theme-secondary, #018786) var(--mdc-slider-tick-width),
      transparent 0)
    0px center /
    calc((100% - var(--mdc-slider-tick-width)) / var(--mdc-slider-tick-count))
    100% repeat-x
  ;
```
This way you don't need to hack a layout with a flex container that has heaps of marker nodes, a negative `margin-left` value, and your `.mdc-slider__track-marker-container:after` pseudo element to offset its end marker.

### PR DESCRIPTION
**Pull request by jonjohnjohnson**
So one of the few components you make it difficult to test for is the slider, but AFAIKT, this completely resolves  #4729. Outside of the suite, I deployed and performed a pretty rigid stress test and it behaviorally matches up. Maybe the remove method could be taken out, since the new append method essentially just replaces values, but I left an updated definition of it just in case it *could be called from elsewhere to hide ticks, between altering tick marker numbers, etc...

cc @moog16 please comb this, even aside from my assurance of it behaving implacably in browser. Meaning, rename "append" to "create" and fix my typo on line 140 of the adapter if you want. :)

### CODE REVIEW

**Comment by jonjohnjohnson** [comment]
I signed it!

**Comment by jonjohnjohnson** [comment]
I'll alter test cases clean this up by the end of the week if you wanna hold off @moog16.

**Comment by moog16** [comment]
Thanks for opening the PR! I will take a look and review. You can hold off on tests until I do a full review so we can reduce rework going back and forth on it. How does that sound?

**Comment by jonjohnjohnson** [comment]
@moog16 cool, figured if it’s a go from ya, we could squash/rebase, pending the linting issue that’s lost on me and chais ability to test the value of a css custom property.

File: packages/mdc-slider/component.ts [hunk 1]
```diff
@@ -117,24 +117,11 @@ export class MDCSlider extends MDCComponent<MDCSliderFoundation> {
...
-        while (this.trackMarkerContainer_.firstChild) {
-          this.trackMarkerContainer_.removeChild(this.trackMarkerContainer_.firstChild);
-        }
-      },
-      setLastTrackMarkersStyleProperty: (propertyName, value) => {
-        // We remove and append new nodes, thus, the last track marker must be dynamically found.
-        const lastTrackMarker = this.root_.querySelector<HTMLElement>(strings.LAST_TRACK_MARKER_SELECTOR)!;
-        lastTrackMarker.style.setProperty(propertyName, value);
+      setTrackMarkers: (step, max, min) => {
+        // TODO iterate on args, but for now explicit to test with
```
Comment (moog16) [inline packages/mdc-slider/component.ts]: I think this is fine without iterating. It's clear.

**Comment by jonjohnjohnson** [comment]
IE11 doesn't support custom properties, but does support `linear-gradient`, so we just have to swap out the each `var()` for lengths and color values in sass. I'll rig in the sass logic and see what I can do for screenshots.

**Comment by moog16** [comment]
Cool please let me know when its ready to be reviewed again?

**Comment by jonjohnjohnson** [comment]
@moog16 mind sharing with me the exact node env you're runnin. No matter what I try, I'm always getting a few different errors, and commonly finding this bad boy...
Screen Shot 2019-06-03 at 11 10 51 PM

**Comment by moog16** [comment]
@jonjohnjohnson sorry I dropped the ball - been out on vacation. I run using 10.13.0. Let me know if you have any more issues. Or if that doesn't fix it.

**Comment by jonjohnjohnson** [comment]
> Secondly I see that you're using CSS Properties here. We want to keep the support of this. Does this still work in IE11? Also we will want a way to customize this with Sass, since all our components work with Sass. Can you please alter your design for it to be customized with Sass like in the other components?

Hey @moog16, the core change to make this approach work is all done through 4dd34a0. Can you just check this branch and `npm run build`, loading in the build to an initialized slider? I don't know if it's worth me monkeying with the test setup aside from you checking out this core change of mine. Unless you'd want to pair in a hangout sometime for a few minutes?

**Comment by moog16** [comment]
What i can do is add a screenshot test into core. This PR will then be dependent on that PR. How does that sound?

**Comment by moog16** [comment]
Created #4816

**Comment by jonjohnjohnson** [comment]
Want me to pull that branch in or just wait until it's merged and see if you still need anything from me?

**Comment by moog16** [comment]
Let's wait until it is merged. We can work from there afterwards.

**Comment by moog16** [comment]
@jonjohnjohnson please pull in master. These changes should not break the screenshots.

**Comment by jonjohnjohnson** [comment]
@moog16 🏓

File: packages/mdc-slider/component.ts [hunk 2]
```diff
@@ -118,24 +118,8 @@ export class MDCSlider extends MDCComponent<MDCSliderFoundation> {
...
-          frag.appendChild(marker);
-        }
-        this.trackMarkerContainer_.appendChild(frag);
-      },
-      removeTrackMarkers: () => {
-        while (this.trackMarkerContainer_.firstChild) {
-          this.trackMarkerContainer_.removeChild(this.trackMarkerContainer_.firstChild);
-        }
-      },
-      setLastTrackMarkersStyleProperty: (propertyName, value) => {
```
Comment (moog16) [inline packages/mdc-slider/component.ts]: Please remove all of these adapter methods from the README.md. Also add the new `setTrackMarkers` method.

File: packages/mdc-slider/component.ts [hunk 3]
```diff
@@ -119,7 +119,16 @@ export class MDCSlider extends MDCComponent<MDCSliderFoundation> {
       setTrackStyleProperty: (propertyName, value) => this.track_.style.setProperty(propertyName, value),
       setMarkerValue: (value) => this.pinValueMarker_.innerText = value.toLocaleString(),
       setTrackMarkers: (step, max, min) => {
-        this.trackMarkerContainer_.style.setProperty('background', 'linear-gradient(to right, currentColor 2px, transparent 0) 0 center / calc((100% - 2px) /  (('+ max.toLocaleString() + ' - ' + min.toLocaleString() + ') / ' + step.toLocaleString() + ')) 100% repeat-x');
+        const maxString = max.toLocaleString();
+        const minString = max.toLocaleString();
+        const stepString = max.toLocaleString();
+        // keep calculation in css for better rounding/subpixel behavior
+        const markerAmount = '((' + maxString + ' - ' + minString + ') / ' + stepString + ')';
```
Comment (moog16) [inline packages/mdc-slider/component.ts]: Can you convert these to using backticks? It might read better.

Comment (jonjohnjohnson) [reply packages/mdc-slider/component.ts]: Oh definitely, I would rather. I just hadn't come across any template literals in the codebase yet, so I wasn't sure if it was your style.

**Comment by jonjohnjohnson** [comment]
@moog16 🏓

**Comment by moog16** [comment]
@jonjohnjohnson yeah the backticks look better. I updated the base branch to develop. Could you please water your branch with `develop`? There were 2 conflicts that I fixed. I think you just need to update.

**Comment by jonjohnjohnson** [comment]
@moog16 honestly not sure if I handled the automerge failure well, tread lightly.

**Comment by moog16** [comment]
i will double check!

**Comment by moog16** [comment]
Looks like there are 46 files changed...originally you only had about 7 files changed. The easiest thing to do if you cannot backtrack is to create a new branch off of `develop`, and cherry-pick your specific commits over. Not the best solution, but it will work. Let me know if you run into issues.

**Comment by jonjohnjohnson** [comment]
I signed it!

**Comment by jonjohnjohnson** [comment]
@moog16 🏓 aside from what I'm guessing is another devs commit to the `develop` branch not passing the google cla, I thiiiiiiink this is fine now???

**Comment by moog16** [comment]
I manually changed the CLA...however the PR still has 39 files changed. Including changes from ripple, dom, base, and others you never touched. So I think the `develop` sync did not go as planned.

Screen Shot 2019-06-19 at 10 01 33 AM

**Comment by jonjohnjohnson** [comment]
@moog16 without sucking a bit of time into this, I have no idea why my attempts to make a clean rebase off of the upstream/develop branch with my changes is still bringing in other commits that aren't already a part of the upstream develop branch. :/

**Comment by jonjohnjohnson** [comment]
@moog16 🏓

File: packages/mdc-slider/component.ts [hunk 4]
```diff
@@ -22,6 +22,7 @@
  */
 
 import {MDCComponent} from '@material/base/component';
+import {applyPassive} from '@material/dom/events';
```
Comment (moog16) [inline packages/mdc-slider/component.ts]: You need to update @material/dom to the package.json of slider from 1.1.0 to the latest.

**Comment by moog16** [comment]
@abhiomkar ready for rereview

**Pull request merged**